k2/csrc/context.cu

```diff
@@ -128,4 +128,50 @@ void GetBlockSizesForLambda2(int32_t m, int32_t n, dim3 *block_dim,
   }
 }
 
+
+void Semaphore:: Signal(ContextPtr c) {
+  DeviceType device_type = c->GetDeviceType();
+  if (device_type_ == kUnk)
+    device_type_ = device_type;
+  else
+    K2_CHECK_EQ(device_type, device_type_)
+        << "Semaphore must always be used with the same device type.";
+  if (device_type == kCuda) {
+    cudaEvent_t event;
+    cudaError_t e = cudaEventCreateWithFlags(&event, cudaEventDisableTiming);
+    K2_CHECK_CUDA_ERROR(e) << "Error creating event";
+    // Note: this stream is subject to being overridden by With(stream..), see
+    // class With.
+    cudaStream_t stream = c->GetCudaStream();
+    e = cudaEventRecord(event, stream);
+    K2_CHECK_CUDA_ERROR(e) << "Error recording event.";
+    std::unique_lock<std::mutex> lock(events_mutex_);
+    events_.push_back(event);
+  }
+  semaphore_.release();
+}
+
+void Semaphore::Wait(ContextPtr c) {
+  DeviceType device_type = c->GetDeviceType();
+  if (device_type_ == kUnk)
+    device_type_ = device_type;
+  else
+    K2_CHECK_EQ(device_type, device_type_)
+        << "Semaphore must always be used with the same device type.";
+  semaphore_.acquire();
```

Comment:
Looks like semaphore_ is used only when the device type is cuda.
Shall we put `acquire` and `release` inside the statement`if (device_type == kCuda) {`
? It can avoid the overhead of using mutex and condition variables when the device type
is cpu, I think.

Reply:
No, the semaphore is needed regardless actually; it's the events that are only needed for CUDA.